addons/nuxeo-core-binarymanager-cloud/nuxeo-core-binarymanager-gcp/src/main/java/org/nuxeo/ecm/core/storage/gcp/GoogleStorageBinaryManager.java

```diff
@@ -0,0 +1,241 @@
+import com.google.cloud.storage.Storage;
+import com.google.cloud.storage.Storage.BlobField;
+import com.google.cloud.storage.Storage.BlobListOption;
+import com.google.cloud.storage.StorageException;
+import com.google.cloud.storage.StorageOptions;
+
+/**
+ * A Binary Manager that stores binaries as Google Storage BLOBs
+ * <p>
+ * The BLOBs are cached locally on first access for efficiency.
+ * <p>
+ * Because the BLOB length can be accessed independently of the binary stream, it is also cached in a simple text file
+ * if accessed before the stream.
+ *
+ * @since 10.10-HF12
+ */
+public class GoogleStorageBinaryManager extends AbstractCloudBinaryManager {
+
+    private static final Log log = LogFactory.getLog(GoogleStorageBinaryManager.class);
+
+    public static final String BUCKET_NAME_PROPERTY = "bucket";
+
+    public static final String PROJECT_ID_PROPERTY = "project";
+
+    public static final String GOOGLE_APPLICATION_CREDENTIALS = "credentials";
+
+    public static final String GOOGLE_PLATFORM_SCOPE = "https://www.googleapis.com/auth/cloud-platform";
+
+    public static final String GOOGLE_STORAGE_SCOPE = "https://www.googleapis.com/auth/devstorage.full_control";
+
+    public static final String SYSTEM_PROPERTY_PREFIX = "nuxeo.gcp";
+
+    protected String bucketName;
+
+    protected Bucket bucket;
+
+    protected Storage storage;
+
+    @Override
+    protected void setupCloudClient() {
+        try {
+            String projectId = getProperty(PROJECT_ID_PROPERTY);
+
+            GoogleCredentials credentials = GoogleCredentials.fromStream(
+                    new ByteArrayInputStream(getProperty(GOOGLE_APPLICATION_CREDENTIALS).getBytes()))
+                                                             .createScoped(GOOGLE_PLATFORM_SCOPE, GOOGLE_STORAGE_SCOPE);
+            credentials.refreshIfExpired();
+
+            storage = StorageOptions.newBuilder()
+                                    .setCredentials(GoogleCredentials.create(credentials.refreshAccessToken()))
+                                    .setProjectId(projectId)
+                                    .build()
+                                    .getService();
+            bucketName = getProperty(BUCKET_NAME_PROPERTY);
+            bucket = getOrCreateBucket(bucketName);
+        } catch (IOException | StorageException e) {
+            throw new NuxeoException(e);
+        }
+    }
+
+    /**
+     * Gets or creates a bucket with the given {@code bucketName}.
+     *
+     * @return the bucket instance.
+     */
+    public Bucket getOrCreateBucket(String bucketName) {
+        try {
+            Bucket bucket = storage.get(bucketName);
+            if (bucket == null) {
+                log.debug(String.format("Creating a new bucket %s", bucketName));
+                return storage.create(BucketInfo.of(bucketName));
+            }
+            return bucket;
+        } catch (StorageException e) {
+            throw new NuxeoException(e);
```

Comment:
`StorageException` is a runtime exception, no need to catch and rethrow it (several occurrences).